table/block_based/block_based_table_builder.cc

```diff
@@ -1501,15 +1501,16 @@ Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
     const size_t read_amp_bytes_per_bit =
         rep_->table_options.read_amp_bytes_per_bit;
 
-    TBlocklike* block_holder = BlocklikeTraits<TBlocklike>::Create(
-        std::move(results), read_amp_bytes_per_bit,
-        rep_->ioptions.statistics.get(),
-        false /*rep_->blocks_definitely_zstd_compressed*/,
-        rep_->table_options.filter_policy.get());
+    std::unique_ptr<TBlocklike> block_holder(
+        BlocklikeTraits<TBlocklike>::Create(
+            std::move(results), read_amp_bytes_per_bit,
+            rep_->ioptions.statistics.get(),
+            false /*rep_->blocks_definitely_zstd_compressed*/,
+            rep_->table_options.filter_policy.get()));
 
     if (block_holder->own_bytes()) {
       size_t charge = block_holder->ApproximateMemoryUsage();
-      s = block_cache->Insert(key, block_holder, charge,
+      s = block_cache->Insert(key, block_holder.release(), charge,
```

Comment:
It is trickier than this because the block cache only takes ownership on OK return. See calls to `block_cache->Insert()` in block_based_table_reader.cc

Reply:
In block_based_table_reader.cc, if ```block_cache_->Insert() ``` returns Not Ok status (failure), it doesn't release the pointer. Does ```insert``` takes care of that?

Reply:
Here we release unique_ptr on OK (skip automatic delete): https://github.com/facebook/rocksdb/blob/dbe3810c74a0103919ccbcccee52feba7e652866/table/block_based/block_based_table_reader.cc#L1202

And here we delete raw ptr on non-OK:
https://github.com/facebook/rocksdb/blob/dbe3810c74a0103919ccbcccee52feba7e652866/table/block_based/block_based_table_reader.cc#L1295

Either works but unique_ptr is obviously preferred.